compiler/src/iree/compiler/Dialect/Util/TransformOps/UtilTransformOps.cpp

```diff
@@ -0,0 +1,328 @@
+                            ? dom.dominates(input, insertionPoint)
+                            : dom.properlyDominates(input, insertionPoint);
+    if (!doesDominate) {
+      return emitDefiniteFailure()
+             << "input " << input << " does not dominate insertion point "
+             << insertionPoint;
+    }
+  }
+
+  // Get the function to call. This can either be specified by symbol or as a
+  // transform handle.
+  FunctionOpInterface targetFunction = nullptr;
+  if (getFunctionName()) {
+    targetFunction = SymbolTable::lookupNearestSymbolFrom<FunctionOpInterface>(
+        insertionPoint, *getFunctionName());
+    if (!targetFunction) {
+      return emitDefiniteFailure()
+             << "unresolved symbol " << *getFunctionName();
+    }
+  } else if (getFunction()) {
+    auto payloadOps = state.getPayloadOps(getFunction());
+    if (!llvm::hasSingleElement(payloadOps)) {
+      return emitDefiniteFailure() << "requires a single function to call";
+    }
+    targetFunction = dyn_cast<FunctionOpInterface>(*payloadOps.begin());
+    if (!targetFunction) {
+      return emitDefiniteFailure() << "invalid non-function callee";
+    }
+  } else {
+    llvm_unreachable("invalid CastAndCall op without a function to call");
+    return emitDefiniteFailure();
+  }
+
+  // Verify that the function argument and result lengths match the inputs and
+  // outputs given to this op.
+  if (targetFunction.getNumArguments() != inputs.size()) {
+    return emitSilenceableFailure(targetFunction.getLoc())
+           << "mismatch between number of function arguments "
+           << targetFunction.getNumArguments() << " and number of inputs "
+           << inputs.size();
+  }
+  if (targetFunction.getNumResults() != outputs.size()) {
+    return emitSilenceableFailure(targetFunction.getLoc())
+           << "mismatch between number of function results "
+           << targetFunction->getNumResults() << " and number of outputs "
+           << outputs.size();
+  }
+
+  // Gather all specified converters.
+  mlir::TypeConverter converter;
+  if (!getRegion().empty()) {
+    for (Operation &op : getRegion().front()) {
+      cast<transform::TypeConverterBuilderOpInterface>(&op)
+          .populateTypeMaterializations(converter);
+    }
+  }
+
+  if (insertAfter)
+    rewriter.setInsertionPointAfter(insertionPoint);
+  else
+    rewriter.setInsertionPoint(insertionPoint);
+
+  for (auto [input, type] :
+       llvm::zip_equal(inputs, targetFunction.getArgumentTypes())) {
+    if (input.getType() != type) {
+      Value newInput = converter.materializeSourceConversion(
+          rewriter, input.getLoc(), type, input);
+      if (!newInput) {
+        return emitDefiniteFailure() << "Failed to materialize conversion of "
+                                     << input << " to type " << type;
+      }
+      input = newInput;
+    }
+  }
+
+  auto callOp = rewriter.create<IREE::Util::CallOp>(insertionPoint->getLoc(),
```

Comment:
Does `util.call` work with `func.func`? Can we either add a helper for `Util::createCall(FunctionOpInterface)` or require a `util.func` above instead of `FunctionOpInterface`?

Reply:
Nope, util.call can only call util.func. Since this is util-specific there's no real need for interfaces as we don't really want this matching any other dialects that may have their own rules around function calls.

Reply:
yeah makes sense, then on line 193 lets require a `Util::FuncOp`: https://github.com/openxla/iree/pull/16426/files#diff-09ef5afa0288aff5180b52a4c380b92c64ee6274d25b19f021870b8a3ea65b31R193

unless there is an intent to allow user defined functions able to be called from a `util.call`, in which case I'd expect we need our own `UtilFunctionOpInterface`.